mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -212,119 +212,123 @@ struct XdlopsGemmV2RewritePattern : public OpConversionPattern<XdlopsGemmV2Op> {
-    //       p_c_thread);
-    // }
-    int64_t outerLoopUpperBound = KPerThread;
-    // In case xdlops consume multiple elements from memref<T>, divide by k_base
-    // such that the generated loop is not out of bounds.
-    if (KPack == 1) {
-      outerLoopUpperBound /= k_base;
-    }
     auto regAConstantOp =
         b.create<ConstantIndexOp>(loc, op.regOffsetAAttr().getInt());
     auto regBConstantOp =
         b.create<ConstantIndexOp>(loc, op.regOffsetBAttr().getInt());
 
-    auto outerLoop = b.create<AffineForOp>(loc, 0, outerLoopUpperBound);
-    auto outerLoopb = ConversionPatternRewriter::atBlockBegin(
-        outerLoop.getBody(), b.getListener());
-    auto outerLoopiv = outerLoop.getInductionVar();
+    auto populateMfma = [&](OpBuilder &b, Value &argA, Value &argB) {
+      for (int64_t i = 0; i < vectorNumber; ++i) {
+        // Note below is assuming only one of MRepeats or NRepeats is larger
+        // than 1, which fits the existing blockwisegemmv2op implementation.
+        // TODO: Move MRepeats and NRepeats into xdlopsgemmv2op
+        int64_t regDOffset = 0;
+        if (op.regOffsetAAttr().getInt() > 0 ||
+            op.regOffsetBAttr().getInt() > 0) {
+          regDOffset += vectorNumber;
+        }
+        Value offset =
+            b.createOrFold<arith::ConstantIndexOp>(loc, regDOffset + i);
+
+        auto vectorC = b.create<memref::LoadOp>(loc, vectorType,
+                                                adaptor.matrixC(), offset);
+        auto mfma = b.create<amdgpu::MFMAOp>(
+            loc, vectorType, mfmaInstr, argA, argB, vectorC,
+            /*cbsz=*/imms[i][0], /*abid=*/imms[i][1], /*blgp=*/imms[i][2]);
+        auto vectorD = mfma.destD();
+
+        b.create<memref::StoreOp>(loc, vectorD, adaptor.matrixC(), offset);
+      }
+    };
 
-    // matrixAElement and matrixBElement are only useful when KPack > 1
-    Value matrixAElement;
-    Value matrixBElement;
+    // TODO: zyin adopt generic layout: K, KRepeats, k_base
+    // After that, we'd be able to uniform between kPack/nonKPack
+    if (KPack == 1) {
+      // for(index_t k_i = 0; k_i < KPerThread; k_i += k_base) {
+      //   argA = a[k_i];
+      //   argB = a[k_i];
+      //   p_c_thread = mfma(argA, argB, p_c_thread);
+      // }
+      Value zeroConstantOp = b.create<ConstantIndexOp>(loc, 0);
+      auto mfmaLoop = b.create<TransformingForOp>(
+          loc, ArrayRef<ValueRange>{{zeroConstantOp}},
+          ArrayRef<Attribute>{b.getArrayAttr({})},
+          /*bounds=*/ArrayRef<int64_t>{KPerThread},
+          /*strides=*/ArrayRef<int64_t>{k_base},
+          /*useIndexDiffs=*/true, /*forceUnroll=*/true);
+      {
+        OpBuilder::InsertionGuard guard(b);
+        b.setInsertionPointToStart(mfmaLoop.getBody());
+        Value coord = mfmaLoop.getLowerCoords(/*domain=*/0)[0];
+
+        Value offset = coord;
+        Value regAIdx = b.create<AddIOp>(loc, regAConstantOp, offset);
```

Comment:
Here's a thought. What if the loop had transforms where we `Slice` to apply the regAOffset and regBOffset, if any? Not sure that'd make it any clearer, but it would lead to interesting things.

We could also have one loop for a and one loop for b, each of which  starts at the relevant register offset.

... in fact, yeah, what if there was a function that gave you this load loop for A or B, as relevant - the only difference between those loops is whether you use `regAOffset` or `regBOffset`. Sure, that'll leave the loads in a different order, but it does mean not needing to copy-paste the A and B logic.

Reply:
Actually I'd prefer instead to make a single loop and call a lambda for a and b each. 

Reasoning being that we know that the right form of code should consolidate both loops into one for max performance. I find it defeat the purpose if we instead make two loops for the hope that compiler would do the right thing for us.

Reply:
Entirely fair. We could have a loop that gives you an a coordinate and a b coordinate.

Although the pattern further up in GEMM is to have a loop for a and a loop for b and it seems to work fine

Reply:
Done in 1405899